### PR TITLE
Refactor URI template to Fix Dispatching Issue #4948

### PR DESCRIPTION
**Pull request by irunika**
This fixes https://github.com/ballerina-lang/ballerina/issues/4948

### CODE REVIEW

File: stdlib/ballerina-http/src/main/java/org/ballerinalang/net/uri/parser/DataElement.java [hunk 1]
```diff
@@ -46,5 +46,5 @@
      * @param inboundMessage Checker for extra checks of the item.
      * @return the item stored in the Node Item.
      */
-    DataType getData(InboundMsgType inboundMessage);
+    boolean getData(InboundMsgType inboundMessage, DataReturnAgent<DataType> dataReturnAgent);
```
Comment (anupama-pathirage) [inline stdlib/ballerina-http/src/main/java/org/ballerinalang/net/uri/parser/DataElement.java]: Change the Doc parameters and return type according to the new change

File: stdlib/ballerina-http/src/main/java/org/ballerinalang/net/uri/parser/DataReturnAgent.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
+package org.ballerinalang.net.uri.parser;
```
Comment (anupama-pathirage) [inline stdlib/ballerina-http/src/main/java/org/ballerinalang/net/uri/parser/DataReturnAgent.java]: Missing License header

File: stdlib/ballerina-http/src/main/java/org/ballerinalang/net/uri/parser/Node.java [hunk 3]
```diff
@@ -59,54 +60,59 @@ public DataElementType getDataElement() {
...
-            return null;
+            return false;
         }
         if (matchLength == uriFragment.length()) {
-            return dataElement;
+            return dataElement.getData(inboundMsg, dataReturnAgent);
         }
         if (matchLength >= uriFragment.length()) {
-            return null;
+            return false;
```
Comment (anupama-pathirage) [inline stdlib/ballerina-http/src/main/java/org/ballerinalang/net/uri/parser/Node.java]: Shall we improve this logic. the = part of matchLength >= uriFragment.length() condition is already satisfied by the matchLength == uriFragment.length(). So this logic is misleading. So shall we re-factor this to something like below which is more clear 
```
int matchLength = match(uriFragment, variables);
int fragmentLength = uriFragment.length();
if (matchLength < 0 || matchLength > fragmentLength ) {
            return false;
        }
        if (matchLength == fragmentLength) {
            return dataElement.getData(inboundMsg, dataReturnAgent);
        }
```

File: stdlib/ballerina-http/src/main/java/org/ballerinalang/net/uri/parser/DataReturnAgent.java (hunk 2 above)
Comment (irunika) [reply stdlib/ballerina-http/src/main/java/org/ballerinalang/net/uri/parser/DataReturnAgent.java]: Done

**Comment by djkevincr** [comment]
Can't we add test case here?

File: stdlib/ballerina-http/src/main/java/org/ballerinalang/net/uri/parser/URITemplateParser.java [hunk 4]
```diff
@@ -34,17 +34,17 @@
...
     private static final char[] operators = new char[] { '+', '.', '/', ';', '?', '&', '#' };
 
-    private Node<DataElement<DataType, InboundMgsType>> syntaxTree;
-    private Node<DataElement<DataType, InboundMgsType>> currentNode;
+    private Node<DataType, InboundMgsType> syntaxTree;
+    private Node<DataType, InboundMgsType> currentNode;
     private final DataElementFactory<? extends DataElement<DataType, InboundMgsType>> elementCreator;
 
-    public URITemplateParser(Node<DataElement<DataType, InboundMgsType>> rootNode,
+    public URITemplateParser(Node<DataType, InboundMgsType> rootNode,
```
Comment (djkevincr) [inline stdlib/ballerina-http/src/main/java/org/ballerinalang/net/uri/parser/URITemplateParser.java]: Can't we add test case?

**Pull request merged**